src/main/java/io/bf2/kafka/authorizer/CustomAclAuthorizer.java

```diff
@@ -246,6 +247,35 @@ void addAllowedListeners(Map<String, ?> configs) {
         }
     }
 
+    /**
+     * To avoid misleading users we delete principals that are explicitly targeted by CustomAclAuthorizer rules.
+     * It is possible for a user with ACL rules in Zookeeper to become a user managed by CustomAclAuthorizer, in which
+     * case we would never use or update rules for them in ZK. The old rulesremain, dangling, in ZK and visible through
+     * kafka ACL tooling. This would mislead users becuase there will be a mismatch between what the ZK rules say they
+     * should be permitted to do and the truth.
+     */
+    private void deleteOrphanBindingsFromDelegate() {
+        List<AclBindingFilter> toDelete = aclPrincipals.stream().map(aclPrincipal -> {
+            AccessControlEntryFilter principalFilter = new AccessControlEntryFilter(aclPrincipal, null, AclOperation.ANY, AclPermissionType.ANY);
+            return new AclBindingFilter(ResourcePatternFilter.ANY, principalFilter);
+        }).collect(Collectors.toList());
+        if (!toDelete.isEmpty()) {
+            log.info("deleting orphaned bindings if present, using filter: {}", toDelete);
+            List<CompletableFuture<AclDeleteResult>> deletionResults = delegate.deleteAcls(null, toDelete)
+                    .stream().map(CompletionStage::toCompletableFuture).collect(Collectors.toList());
+            CompletableFuture<Void> future = CompletableFuture.allOf(deletionResults.toArray(CompletableFuture[]::new));
+            future.whenComplete((unused, throwable) -> {
```

Comment:
Have gone for a single fire-and-forget deletion attempt that should not block startup.